# Regression in Python

In [10]:
path='PRSA_data_2010.1.1-2014.12.31.csv'
f = open(path, 'r')

In [11]:
dataset = []
header = f.readline().strip().split(',')
for line in f:
    line = line.split(',')
    dataset.append(line)

In [12]:
N = len(dataset)
N

43824

In [13]:
header

['No',
 'year',
 'month',
 'day',
 'hour',
 'pm2.5',
 'DEWP',
 'TEMP',
 'PRES',
 'cbwd',
 'Iws',
 'Is',
 'Ir']

In [14]:
header.index('pm2.5')

5

In [15]:
header.index('TEMP')

7

In [16]:
y = [float(d[5]) for d in dataset]

ValueError: could not convert string to float: 'NA'

In [17]:
dataset = [d for d in dataset if d[5] != 'NA']

In [18]:
y = [float(d[5]) for d in dataset]

In [19]:
def feature(datum):
    feat = [1, float(datum[7])]
    return feat

In [20]:
X = [feature(d) for d in dataset]

In [21]:
y[:10]

[129.0, 148.0, 159.0, 181.0, 138.0, 109.0, 105.0, 124.0, 120.0, 132.0]

In [22]:
X[:10]

[[1, -4.0],
 [1, -4.0],
 [1, -5.0],
 [1, -5.0],
 [1, -5.0],
 [1, -6.0],
 [1, -6.0],
 [1, -5.0],
 [1, -6.0],
 [1, -5.0]]

# Why did we implement our feature function like this?

def feature(datum):
    feat = [1, float(datum[7])]
    return feat

"""
To help solve the equation pm2.5 = theta0 + theta1(temp)
this can be expressed as: (theta0,theta1)*(1, temp)
which is a dot product

"""

In [23]:
import numpy
theta, residuals,rank,s = numpy.linalg.lstsq(X,y)

<ipython-input-23-af50c4dbc7d9>:2: FutureWarning: `rcond` parameter will change to the default of machine precision times ``max(M, N)`` where M and N are the input matrix dimensions.
To use the future default and silence this warning we advise to pass `rcond=None`, to keep using the old, explicitly pass `rcond=-1`.
  theta, residuals,rank,s = numpy.linalg.lstsq(X,y)


In [24]:
theta

array([107.10183392,  -0.68447989])

^ This gives us that the line of best fit is:
pm2.5 = 107.1 -0.68*temp

In [ ]:
# we can change the output by adding features to the feature function
# we added pressure and wind speed
def feature(datum):
    feat = [1, float(datum[7]), float(datum[8]), float(datum[10])]
    return feat

In [ ]:
X = [feature(d) for d in dataset]
theta, residuals,rank,s = numpy.linalg.lstsq(X,y)
theta

The new equation would be:
pm2.5 = 3263.7 - 3.109*temp - 3.065*pressure - 0.460*wind speed

In [ ]:
# Lets do this the brute force way instead of using a library to see if the result is the same
X = numpy.matrix(X)
y = numpy.matrix(y)
numpy.linalg.inv(X.T * X) * X.T * y.T

# Time-series regression

we would like to predict a sequence of real-valued events as accurately as possible 

# Autoregression

Can we predict the next air quality measurement from the previous ones

In [ ]:
import numpy

In [ ]:
path='PRSA_data_2010.1.1-2014.12.31.csv'
f = open(path, 'r')

In [ ]:
dataset = []
header = f.readline().strip().split(',')
for line in f:
    line = line.split(',')
    dataset.append(line)

In [ ]:
dataset = [d for d in dataset if d[5] != 'NA']

In [ ]:
def feature(dataset, ind, windowSize):
    feat = [1]
    previousValues = [float(d[5]) for d in dataset[ind-windowSize:ind]]
    return feat + previousValues

In [ ]:
windowSize = 10
N = len(dataset)

In [ ]:
X = [feature(dataset, ind, windowSize) for ind in range(windowSize, N)]

In [ ]:
X[:10]

In [ ]:
y = [float(d[5]) for d in dataset[windowSize:]]

In [ ]:
theta, resuduals,rank,s = numpy.linalg.lstsq(X,y)

In [ ]:
# listed from oldest to most recent
theta

In [ ]:
# slightly more sofisticated example of regression

def feature(dataset, ind, windowSize):
    feat = [1, float(dataset[ind][7]), float(dataset[ind][8]), float(dataset[ind][10])]
    previousValues = [float(d[5]) for d in dataset[ind-windowSize:ind]]
    return feat + previousValues

In [ ]:
X = [feature(dataset, ind, windowSize) for ind in range(windowSize, N)]

In [ ]:
theta, resuduals,rank,s = numpy.linalg.lstsq(X,y)

In [ ]:
theta

# Features from Categorical Data

Working with missing values

We will look at a few different options fot handle the missing data

    - Filtering
    - Missing data imputation: try to fill in data with reasonable estimates
    - Modeling: change algerithum to handle missing data

Missing data imputation

A simple scheme would be to replace every missing values with the average.
    - This could cause problems if you have large outliers.
    - The imputed value may not be realistic

An alternative would be to consider using subgroups

We could also train a separate predictor
    - This could be a problem if this data is missing also

Add an additional feature to indicate 'missing feature'

Classification in Python

In [ ]:
f = open('bankruptcy/5year.arff', 'r')

In [ ]:
# the header ends and the real data begins after we see the '@data' tag
while not '@data' in f.readline():
    pass

In [ ]:
dataset = []
for l in f:
    if '?' in l:
        continue
    l = l.split(',')
    values = [1] + [float(x) for x in l]
    values[-1] = values[-1] > 0 # Convert to bool
    dataset.append(values)

In [ ]:
len(dataset)

In [ ]:
sum([x[-1] for x in dataset])

In [ ]:
X = [values[:-1] for values in dataset]

In [ ]:
Y = [values[-1] for values in dataset]

In [ ]:
from sklearn import linear_model

In [ ]:
model = linear_model.LogisticRegression()

In [ ]:
model.fit(X,Y)

In [ ]:
predictions = model.predict(X)

In [ ]:
predictions

In [ ]:
# find which predictions were correct
correctPredictions = predictions == Y

In [ ]:
correctPredictions

In [ ]:
# find accuracy of predictions
sum(correctPredictions) / len(correctPredictions)

Training and testing data

In [ ]:
f = open('bankruptcy/5year.arff', 'r')

In [ ]:
# the header ends and the real data begins after we see the '@data' tag
while not '@data' in f.readline():
    pass

In [ ]:
dataset = []
for l in f:
    if '?' in l:
        continue
    l = l.split(',')
    values = [1] + [float(x) for x in l]
    values[-1] = values[-1] > 0 # Convert to bool
    dataset.append(values)

In [ ]:
import random

In [ ]:
random.shuffle(dataset)

In [ ]:
X = [values[:-1] for values in dataset]

In [ ]:
y = [values[-1] for values in dataset]

In [ ]:
N = len(X)
X_train = X[:N//2]
X_test = X[N//2:]
y_train = y[:N//2]
y_test = y[N//2:]

In [ ]:
len(X), len(X_train), len(X_test)

In [ ]:
from sklearn import linear_model

In [ ]:
model = linear_model.LogisticRegression()

In [ ]:
model.fit(X_train, y_train)

In [ ]:
predictionsTrain = model.predict(X_train)
predictionsTest = model.predict(X_test)

In [ ]:
correctPredictionsTrain = predictionsTrain == y_train
correctPredictionsTest = predictionsTest == y_test

In [ ]:
sum(correctPredictionsTrain)/ len(correctPredictionsTrain)

In [ ]:
sum(correctPredictionsTest)/ len(correctPredictionsTest)

# Gradient Descent

In [54]:
path='PRSA_data_2010.1.1-2014.12.31.csv'
f = open(path, 'r')

In [55]:
dataset = []
header = f.readline().strip().split(',')
for line in f:
    line = line.split(',')
    dataset.append(line)

In [56]:
header.index('pm2.5')

5

In [57]:
dataset = [d for d in dataset if d[5] != 'NA']

In [58]:
def feature(datum):
    feat = [1, float(datum[7])] # temperature
    return feat

In [59]:
X = [feature(d) for d in dataset]
y = [float(d[5]) for d in dataset]

In [60]:
X[0]

[1, -4.0]

In [61]:
K = len(X[0])
K

2

In [62]:
theta = [0.0]*K

In [63]:
theta[0] = sum(y) / len(y)

In [64]:
# function for the inner product
def inner(x,y):
    return sum([a*b for (a,b) in zip(x,y)])

In [65]:
# function for the normal of a vector
def norm(x):
    return sum([a*a for a in x]) # equivalently, inner(x,x)

In [66]:
# function to find the partial derivative
# MSE is Mean Square error
def derivative(X,y, theta):
    dtheta = [0.0]*len(theta)
    K = len(theta)
    N = len(X)
    MSE = 0
    for i in range(N):
        error = inner(X[i],theta) - y[i]
        for k in range(K):
            dtheta[k] += 2*X[i][k]*error/N
        MSE += error*error/N
    return dtheta, MSE

In [67]:
learningRate = 0.003

In [68]:
while(True):
    dtheta,MSE = derivative(X, y, theta)
    m = norm(dtheta)
    print('norm(dtheta) = ' + str(m) + ' MSE = ' + str(MSE))
    for k in range(K):
        theta[k] -= learningRate * dtheta[k]
    if m < 0.01: 
        break

norm(dtheta) = 41178.18180276439 MSE = 8473.070863045637
norm(dtheta) = 27391.037310856456 MSE = 8461.470120068407
norm(dtheta) = 18227.59616521803 MSE = 8453.69132148803
norm(dtheta) = 12137.19327414051 MSE = 8448.453092769803
norm(dtheta) = 8089.214952690785 MSE = 8444.903795606504
norm(dtheta) = 5398.687652667358 MSE = 8442.47740594691
norm(dtheta) = 3610.359546776664 MSE = 8440.797726662375
norm(dtheta) = 2421.658035196417 MSE = 8439.614724709467
norm(dtheta) = 1631.4852346051546 MSE = 8438.762219100408
norm(dtheta) = 1106.18608452748 MSE = 8438.12975918841
norm(dtheta) = 756.9298365044901 MSE = 8437.643934223193
norm(dtheta) = 524.6771841240179 MSE = 8437.255951177423
norm(dtheta) = 370.1890137650022 MSE = 8436.933378638194
norm(dtheta) = 267.3859519156493 MSE = 8436.65465957273
norm(dtheta) = 198.93494191748107 MSE = 8436.40546436858
norm(dtheta) = 153.3159642979812 MSE = 8436.176266962746
norm(dtheta) = 122.87250029772292 MSE = 8435.960733873406
norm(dtheta) = 102.5158175537563 

In [ ]:
theta

Gradient descent with TensorFlow

In [36]:
import tensorflow as tf

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:523: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:524: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/

In [37]:
path='PRSA_data_2010.1.1-2014.12.31.csv'
f = open(path, 'r')

In [38]:
dataset = []
header = f.readline().strip().split(',')
for line in f:
    line = line.split(',')
    dataset.append(line)

In [39]:
header.index('pm2.5')

5

In [40]:
dataset = [d for d in dataset if d[5] != 'NA']

In [41]:
def feature(datum):
    feat = [1, float(datum[7]), float(datum[8]), float(datum[10])]  # temperature, pressure, and wind speed
    return feat

In [42]:
X = [feature(d) for d in dataset]
y = [float(d[5]) for d in dataset]

In [43]:
y = tf.constant(y, shape=[len(y),1])

In [44]:
K = len(X[0])

In [45]:
def MSE(X, y, theta):
    return tf.reduce_mean((tf.matmul(X,theta) - y)**2)

In [46]:
theta = tf.Variable(tf.constant([0.0]*K, shape=[K,1]))

In [47]:
# stochastic gradient descent optimizer with learning rate of 0.01
optimizer = tf.train.AdamOptimizer(0.01)

In [48]:
objective = MSE(X,y,theta)

In [49]:
train = optimizer.minimize(objective)

In [50]:
init = tf.global_variables_initializer()

In [51]:
sess = tf.Session()
sess.run(init)

In [52]:
for iteration in range(1000):
    cvalues = sess.run([train, objective])
    print("objective = " + str(cvalues[1]))

objective = 18197.11
objective = 16256.274
objective = 14544.915
objective = 13065.931
objective = 11818.9375
objective = 10800.008
objective = 10000.627
objective = 9406.955
objective = 8999.841
objective = 8755.391
objective = 8645.959
objective = 8641.206
objective = 8709.448
objective = 8819.531
objective = 8943.106
objective = 9056.544
objective = 9142.379
objective = 9189.911
objective = 9194.929
objective = 9158.805
objective = 9087.197
objective = 8988.592
objective = 8872.917
objective = 8750.339
objective = 8630.25
objective = 8520.496
objective = 8426.892
objective = 8352.981
objective = 8299.985
objective = 8267.004
objective = 8251.36
objective = 8249.125
objective = 8255.695
objective = 8266.347
objective = 8276.795
objective = 8283.601
objective = 8284.418
objective = 8278.101
objective = 8264.656
objective = 8245.025
objective = 8220.871
objective = 8194.206
objective = 8167.128
objective = 8141.5273
objective = 8118.9033
objective = 8100.2065
objective = 8085.7983
obje

In [53]:
with sess.as_default():
    print(MSE(X,y,theta).eval())
    print(theta.eval())

7836.5093
[[ 0.23223479]
 [-0.89481604]
 [ 0.11925128]
 [-0.4959688 ]]


Live coding with TensorFlow

In [1]:
path='PRSA_data_2010.1.1-2014.12.31.csv'
f = open(path, 'r')

In [2]:
dataset = []
header = f.readline().strip().split(',')
for line in f:
    line = line.split(',')
    dataset.append(line)

In [3]:
header.index('pm2.5')

5

In [4]:
dataset = [d for d in dataset if d[5] != 'NA']

In [5]:
def feature(datum):
    feat = [1, float(datum[7]), float(datum[8]), float(datum[10])]
    return feat

In [6]:
X = [feature(d) for d in dataset]
y = [float(d[5]) for d in dataset]
y_length = len(y)

In [7]:
import tensorflow as tf

/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:523: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint8 = np.dtype([("qint8", np.int8, 1)])
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:524: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_quint8 = np.dtype([("quint8", np.uint8, 1)])
/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/tensorflow/python/framework/dtypes.py:525: FutureWarning: Passing (type, 1) or '1type' as a synonym of type is deprecated; in a future version of numpy, it will be understood as (type, (1,)) / '(1,)type'.
  _np_qint16 = np.dtype([("qint16", np.int16, 1)])
/

In [8]:
# tf.constant means these values are constants and not values to be learned
# shape changes the values to a column vector instead of a row vector
y = tf.constant(y, shape=[y_length,1])

In [9]:
K = len(X[0])

In [10]:
def MSE(X,y,theta):
    return tf.reduce_mean((tf.matmul(X,theta) - y)**2)

In [11]:
theta = tf.Variable(tf.constant([0.0]*K, shape=[K,1]))

In [12]:
optimizer = tf.train.AdamOptimizer(0.01)

In [13]:
objective = MSE(X,y,theta)

In [14]:
train = optimizer.minimize(objective)

In [15]:
init = tf.global_variables_initializer()

In [16]:
sess = tf.Session()
sess.run(init)

In [17]:
for iteratio in range(1000):
    cvalues = sess.run([train,objective])
    print('objective = ' + str(cvalues[1]))

objective = 18197.11
objective = 16256.274
objective = 14544.915
objective = 13065.931
objective = 11818.9375
objective = 10800.008
objective = 10000.627
objective = 9406.955
objective = 8999.841
objective = 8755.391
objective = 8645.959
objective = 8641.206
objective = 8709.448
objective = 8819.531
objective = 8943.106
objective = 9056.544
objective = 9142.379
objective = 9189.911
objective = 9194.929
objective = 9158.805
objective = 9087.197
objective = 8988.592
objective = 8872.917
objective = 8750.339
objective = 8630.25
objective = 8520.496
objective = 8426.892
objective = 8352.981
objective = 8299.985
objective = 8267.004
objective = 8251.36
objective = 8249.125
objective = 8255.695
objective = 8266.347
objective = 8276.795
objective = 8283.601
objective = 8284.418
objective = 8278.101
objective = 8264.656
objective = 8245.025
objective = 8220.871
objective = 8194.206
objective = 8167.128
objective = 8141.5273
objective = 8118.9033
objective = 8100.2065
objective = 8085.7983
obje

In [18]:
with sess.as_default():
    print(MSE(X, y, theta).eval())
    print(theta.eval())

7836.5093
[[ 0.23223479]
 [-0.89481604]
 [ 0.11925128]
 [-0.4959688 ]]
